Fix correct warning with c engine when skipping lines

pankajp opened the pull request:
Fixed bug where c engine would not print warnings for lines
it skipped in case the skipped line had an inline comment.
Also, its accounting of number of fields in such lines would
be off by one.

 - [x] closes #16472 
 - [x] tests added
 - [x] tests passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

Commit:
Fix correct warning with c engine when skipping lines

Fixed bug where c engine would not print warnings for lines
it skipped in case the skipped line had an inline comment.
Also, its accounting of number of fields in such lines would
be off by one.

TomAugspurger commented on pandas/tests/io/parser/comment.py:
```diff
@@ -116,3 +118,29 @@ def test_commment_first_line(self):
         result = self.read_csv(StringIO(data), comment='#', header=None)
         tm.assert_frame_equal(result, expected)
+
+    def test_comment_whitespace_delimited(self):
```
You can use the `tm.capture_stderr` decorator from `pandas/util/testing.py` so you don't have to do the patching yourself. See https://github.com/pandas-dev/pandas/blob/6614e266bca32771c761e904367eff10dd4c8979/pandas/tests/io/parser/test_textreader.py#L146 for an example.

TomAugspurger reviewed (changes requested):
Thanks. Can you also add a release note in `doc/source/whatsnew/0.20.2.txt` in the bug fixes section.

jreback commented:
does this have an issue number?

Commit:
Use `tm.capture_stderr` to capture stderr

Commit:
Add bug fix note in `whatsnew/v0.20.3.txt`

Commit:
Move test to CParserTests

The behavior is only applicable on the `c` engine.

pankajp commented:
@TomAugspurger I have added the whatsnew bug entry and replaced `sys.stderr` patching with `tm.capture_stderr`, and also moved the test to `CParserTests` since it is only applicable to the `c` engine.
The CI tests seem to fail on some pyarrow/feather code.

@jreback Added an issue for this: https://github.com/pandas-dev/pandas/issues/16472

jreback commented:
can you rebase on master, CI has been fixed

jreback commented on doc/source/whatsnew/v0.20.2.txt:
```diff
@@ -53,6 +53,7 @@ Indexing
 I/O
 ^^^
 
+- Bug in skipping error lines with inline comments in space delimited text files (:issue:`16472`)
```
something like this: Bug in ``pd.read_csv()`` when ``comment`` is passed in space deliminted text files.

jreback reviewed (approved):
lgtm, minor comment.

jreback commented:
@gfyoung if you can have a look

Commit:
Update whatsnew bug entry as per review

gfyoung commented:
LGTM!

TomAugspurger commented:
Thanks @pankajp!
